### PR TITLE
Introduction of an AudioServlet to provide audio streams through HTTP

### PR DESCRIPTION
**Pull request by kaikreuzer**
Signed-off-by: Kai Kreuzer <kai@openhab.org>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        String streamId = StringUtils.substringAfterLast(req.getRequestURI(), "/");
+        if (!streams.containsKey(streamId)) {
+            logger.debug("Received request for invalid stream id at  {}", req.getRequestURI());
+            resp.sendError(HttpServletResponse.SC_NOT_FOUND);
+        } else {
+            logger.debug("Stream to serve is {}", streamId);
+
+            AudioStream stream = streams.get(streamId);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java]: Shouldn't we use `streams.remove(streamId)` here?

Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java]: Definitely yes! Good catch, I only changed that while debugging and missed to replace it again.

Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java]: fixed.

**Comment by kgoderis** [comment]
@kaikreuzer Kai, what is the rationale behind the fact that a stream will only be served a single time?

**Comment by kaikreuzer** [comment]
Security, Resource-cleanup, etc.

Note that I am aware that this does not work for Sonos as they always do two requests.
I plan to introduce further methods like `serve(AudioStream stream, int times, int maxSeconds)` or something like that to also cover cases like Sonos. Makes sense?

**Comment by kgoderis** [comment]
So far, with Sonos, in the PoC, I have taken the route that each player encodes a wav file to be served. If we play or stream or TTS to a bunch of Sonos at the same time, we effectively have a bunch of identical files (apart from their temp. filenames) (well, Audiostreams of course) to be served 2 times. It could be an optimisation, when/if we allow wildcards to be used in play(..) in Rules - and in the context of audioservlet - to generate/copy over the wav only once but stream it many times. Or, foresee a mechanism that an Audiostream can be served to a player that is not necessarily linked to the ThingHandler that made the original serve() call. Maybe it was your intention to *defacto* serve any AudioStream, and somehow embed an URL in the object, so that URL-driven players can simply pick up the URL on the servlet without further processing or copying to be done? e.g. All Audiostreams get  served whether we want it or not, and when passing on the AudioStream to the Sink, let the Sink decide wether to take the AudioStream or directly move on with the URL that is provided

 I am only afraid that when you have a large number of Sinks that need to play the same stream, that because of encoding or other delays, the streams will not be played in sync. Ok, Sonos has the ZoneGroup stuff, but that is not necessarily true for other brands of players. 

Just thinking out loud.

**Comment by kaikreuzer** [comment]
The AudioSink has to deal with any kind of AudioStream that is passed to it, so we cannot assume that it is a file-based stream. Writing any stream to disk just to be potentially able to serve it multiple times imho does not make much sense. Keeping a copy of the whole stream in memory doesn't make sense either. My current approach would be to use "resettable" streams, so that the same stream can be served multiple times (and it is up to the implementation of the stream whether it holds the data in memory or on the file system or anywhere else). I noticed that resettable might cause concurrency issues though, so it might be better to introduce something like stream cloning without duplicating the data (does anything like this exist in Java already...?).

You can get an idea of how Sonos AudioSink support may look like in my branch at https://github.com/kaikreuzer/smarthome/commit/f75b78192917b567848729824aa6c37adcee9588. Feel free to comment.

**Pull request merged**